### PR TITLE
SC: add p2p-based remote backend networking

### PR DESCRIPTION
**Pull request by jeonghonah**
## Proposed changes

- In order to improve remote backend networking, existing websocket-based code is removed and p2p-based rpc networking is introduced.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- https://github.com/ground-x/klaytn/issues/3725

## Further comments

- passes `klaytn-node-test`.
- includes `dev-service-chain` commits

### CODE REVIEW

File: node/sc/main_bridge_handler.go [hunk 1]
```diff
@@ -75,6 +82,32 @@ func (mbh *MainBridgeHandler) HandleSubMsg(p BridgePeer, msg p2p.Msg) error {
 	return nil
 }
 
+func (mbh *MainBridgeHandler) handleCallMsg(p BridgePeer, msg p2p.Msg) error {
+	// Decode p2p message
+	// TODO-Klaytn-ServiceChain: check msg.Size is enough.
```
Comment (KimKyungup) [inline node/sc/main_bridge_handler.go]: Is this TODO is different one with below ErrRPCBufferOverflow?

File: node/sc/main_bridge_handler.go [hunk 2]
```diff
@@ -75,6 +82,32 @@ func (mbh *MainBridgeHandler) HandleSubMsg(p BridgePeer, msg p2p.Msg) error {
...
+	logger.Trace("mainbridge writes the rpc call message to rpc server", "msg.Size", msg.Size, "msg", msg)
+	data := make([]byte, msg.Size)
+	err := msg.Decode(&data)
+	if err != nil {
+		logger.Error("error in mainbridge message handler", err)
+		return err
+	}
+	if uint32(len(data)) > msg.Size {
```
Comment (KimKyungup) [inline node/sc/main_bridge_handler.go]: As we discussed, let's remove this. :)

File: node/sc/main_bridge_handler.go [hunk 3]
```diff
@@ -75,6 +82,32 @@ func (mbh *MainBridgeHandler) HandleSubMsg(p BridgePeer, msg p2p.Msg) error {
...
 }
 
+func (mbh *MainBridgeHandler) handleCallMsg(p BridgePeer, msg p2p.Msg) error {
+	// Decode p2p message
+	// TODO-Klaytn-ServiceChain: check msg.Size is enough.
+	logger.Trace("mainbridge writes the rpc call message to rpc server", "msg.Size", msg.Size, "msg", msg)
+	data := make([]byte, msg.Size)
+	err := msg.Decode(&data)
+	if err != nil {
+		logger.Error("error in mainbridge message handler", err)
```
Comment (KimKyungup) [inline node/sc/main_bridge_handler.go]:
```
err -> "err", err
```

File: node/sc/main_bridge_handler.go [hunk 4]
```diff
@@ -75,6 +82,32 @@ func (mbh *MainBridgeHandler) HandleSubMsg(p BridgePeer, msg p2p.Msg) error {
...
+		logger.Error("received mainbridge rpc call message is bigger than buffer", "msg", msg.Size, "data", len(data))
+		return ErrRPCBufferOverflow
+	}
+
+	logger.Trace("mainbridge writes the rpc call message to rpc server", "data", len(data))
+
+	// Write to RPC server pipe
+	_, err = mbh.mainbridge.rpcConn.Write(data)
+	if err != nil {
+		logger.Error("write to RPC server failed", err)
```
Comment (KimKyungup) [inline node/sc/main_bridge_handler.go]:
```
err -> "err", err
```

File: node/sc/mainbridge.go [hunk 5]
```diff
@@ -368,6 +406,21 @@ func (pm *MainBridge) handle(p BridgePeer) error {
...
+	logger.Trace("mainbridge send rpc response data to peers", "data len", len(data), "peers", len(peers))
+	for _, peer := range peers {
+		err := peer.SendResponseRPC(data)
+		if err != nil {
+			logger.Error("failed to send rpc response to the peer", "err", err)
+		}
+		return err
+	}
+
+	// TODO-Klaytn Added timeout basedon ctx
```
Comment (KimKyungup) [inline node/sc/mainbridge.go]: What is this context in TODO?

Comment (jeonghonah) [reply node/sc/mainbridge.go]: Removed!

File: node/sc/remote_backend.go [hunk 6]
```diff
@@ -20,147 +20,233 @@ import (
...
+	if err == nil {
+		if r == nil {
+			return nil, klaytn.NotFound
+		}
+	}
+	return r, err
 }
 
-// ChainID can return the chain ID of the chain.
+// ChainID can return the chain ID of the chain by using the p2p client.
```
Comment (KimKyungup) [inline node/sc/remote_backend.go]: Need to change comment. :)

File: node/sc/subbridge.go [hunk 7]
```diff
@@ -297,11 +331,12 @@ func (sc *SubBridge) SetComponents(components []interface{}) {
 	sc.requestEventSub = sc.bridgeManager.SubscribeTokenReceived(sc.requestEventCh)
 	sc.handleEventSub = sc.bridgeManager.SubscribeTokenWithDraw(sc.handleEventCh)
 
-	if err := sc.bridgeManager.RestoreBridges(); err != nil {
-		logger.Error("failed to sc.bridgeManager.RestoreBridges()", "err", err)
-	} else {
-		sc.isAllBridgeRestored = true
-	}
+	//if err := sc.bridgeManager.RestoreBridges(); err != nil {
```
Comment (KimKyungup) [inline node/sc/subbridge.go]: Is it ok to remove this commented code? :)

File: node/sc/subbridge.go [hunk 8]
```diff
@@ -206,6 +209,37 @@ func NewSubBridge(ctx *node.ServiceContext, config *SCConfig) (*SubBridge, error
 	return sc, nil
 }
 
+func (sb *SubBridge) SetRPCConn(conn net.Conn) {
+	sb.rpcConn = conn
+	go func() {
+		for {
+			data := make([]byte, rpcBufferSize)
```
Comment (KimKyungup) [inline node/sc/subbridge.go]: How about add quit routine by `<-sc.quitSync:`?

File: node/sc/subbridge.go [hunk 9]
```diff
@@ -483,15 +513,32 @@ func (pm *SubBridge) handle(p BridgePeer) error {
...
 	defer sc.pmwg.Done()
 
-	report := time.NewTicker(1 * time.Second)
-	defer report.Stop()
-
 	// Keep waiting for and reacting to the various events
 	for {
 		select {
+		case sendData := <-sc.rpcSendCh:
+			sc.SendRPCData(sendData)
```
Comment (KimKyungup) [inline node/sc/subbridge.go]: How about `go sc.SendRPCData(sendData)`?

File: node/sc/subbridge.go (hunk 8 above)
Comment (jeonghonah) [reply node/sc/subbridge.go]: Update error handling for the rpc client. PTAL.

**Pull request merged**